docs/en/lua-developer-faq.md

```diff
@@ -64,108 +88,143 @@ node.restart(); for i = 1, 20 do print("not quite yet -- ",i); end
+```Lua
+   for k,v in pairs(_G) do print(k,v) end
+```
+*  Local variables are 'lexically scoped', and you may declare any variables as local within nested blocks or functions without affecting the enclosing scope.
+*  Because locals are lexically scoped you can also refer to local variables in an outer scope and these are still accessible within the inner scope. Such variables are know as **upvalues**..
+*  Lua variable can be assigned two broad types of data: **values** such as numbers, booleans, and strings and **references** such as functions, tables and userdata.  You can see the difference here  when you assign the contents of a variable `a` to `b`.  In the case of a value then it is simply copied into `b`.  In the case of a reference, both `a` and `b` now refer to the *same object*, and no copying of content takes place.  This process of referencing can have some counter-intuitive consequences. For example, in the following code by the time it exists, the variable `timer2func` is out of scope.  However a reference to the function has now been stored in the Lua registry by the alarm API call, so it and any upvalues that it uses will persist until it is eventually entirely dereferenced (e.g. by `tmr2:unregister()`.
+```Lua 
+  do
+    local tmr2unc = function() ds.convert_T(true) tmr1:start() end
+    tmr2:alarm(300000, tmr.ALARM_AUTO, tmr2unc)
+  end
+  -- 
+```
+*  You need to understand the difference between when a function is compiled, when it is bound as a closure and when it is invoked at runtime. The closure is normally bound once pretty much immediately after compile, but this isn't necessarily the case.  Consider the following example from my MCP23008 module below.
+```Lua 
+  --  Bind the read and write functions for commonly accessed registers
+
+  for reg, regAddr in pairs {
+    IODOR = 0x00,
+    GPPU  = 0x06,  -- Pull-up resistors register for MCP23008
+    GPIO  = 0x09,
+    OLAT = 0x0A,
+    } do  
+    dev['write'..reg] = function(o, dataByte)
+                          write(MCP23008addr, regAddr, dataByte)
+                        end
+    dev['read'..reg]  = function(o)
+                          return read(MCP23008addr, regAddr)
+                        end
+  end
+``` 
+*  This loop is compiled once when the module is required.   Th opcode vectors for the read and write functions are created during the compile, along with a header which defines how many upvalues and locals are used by each function. However, these two functions are then bound _four_ times as different functions (e.g. `mcp23008.writeIODOR()`) and each closure inherits it's own copies of the upvalues it uses so the `regAddr` for this function is `0x00`). The upvalue list is created when the closure is created and though some Lua magic, even if the outer routine which initially declared then is no longer in scope and is GCed (Garbage Collected), the Lua RTS ensures that upvalue still persist whilst the closure persists.
```

Comment:
I agree that using locals will in general generate less code and execute faster, but that is a different point than the issue about locals used as upvalues inside functions will still persist even though the binding function is not longer scoped. 

The second example is complicated but it does underline the basic point that a function is compiled once but can be bound many times at run time.